InfrastructureManager/agents/azure_agent.py

```diff
@@ -0,0 +1,454 @@
+        storage_profile=StorageProfile(os_disk=os_disk)))
+
+    # Sleep until an IP address gets associated with the VM.
+    while True:
+      public_ip_address = network_client.public_ip_addresses.get(resource_group,
+                                                                 vm_network_name)
+      if public_ip_address.ip_address:
+        utils.log('Azure VM is available at {}'.
+                  format(public_ip_address.ip_address))
+        break
+      utils.log("Waiting {} second(s) for IP address to be available".
+                format(self.SLEEP_TIME))
+      time.sleep(self.SLEEP_TIME)
+
+  def associate_static_ip(self, instance_id, static_ip):
+    """Associates the given static IP address with the given instance ID.
+    Args:
+      instance_id: A str that names the instance that the static IP should be
+        bound to.
+      static_ip: A str naming the static IP to bind to the given instance.
+    """
+
+  def terminate_instances(self, parameters):
+    """ Deletes the instances specified in 'parameters' running in Azure.
+    Args:
+      parameters: A dict, containing all the parameters necessary to
+        authenticate this user with Azure.
+    """
+    credentials = self.open_connection(parameters)
+    resource_group = parameters[self.PARAM_RESOURCE_GROUP]
+    subscription_id = str(parameters[self.PARAM_SUBCR_ID])
+    public_ips, private_ips, instance_ids = self.describe_instances(parameters)
+
+    utils.log("Terminating the vm instance/s '{}'".format(instance_ids))
+    compute_client = ComputeManagementClient(credentials, subscription_id)
+    for vm_name in instance_ids:
+      result = compute_client.virtual_machines.delete(resource_group, vm_name)
+      resource_name = 'Virtual Machine' + ':' + vm_name
+      self.sleep_until_delete_operation_done(result, resource_name)
+
+  def sleep_until_delete_operation_done(self, result, resource_name):
+    """ Sleeps until the delete operation for the resource is completed
+    successfully.
+    Args:
+      result: An instance, of the AzureOperationPoller to poll for the status
+        of the operation being performed.
+      resource_name: The name of the resource being deleted.
+    """
+    while not result.done():
+      utils.log("Waiting {0} second(s) for '{1}' to be "
+                "deleted".format(self.SLEEP_TIME, resource_name))
+      time.sleep(self.SLEEP_TIME)
+
+  def cleanup_state(self, parameters):
+    """ Removes any remote state that was created to run AppScale instances
+    during this deployment.
+    Args:
+      parameters: A dict that includes keys indicating the remote state
+        that should be deleted.
+    """
+    subscription_id = str(parameters[self.PARAM_SUBCR_ID])
+    resource_group = parameters[self.PARAM_RESOURCE_GROUP]
+    credentials = self.open_connection(parameters)
+    network_client = NetworkManagementClient(credentials, subscription_id)
+
+    utils.log("Deleting the Virtual Network, Public IP Address "
```

Comment:
Yes, the operations need to be sequential because Network Interface uses the Public IP Address which uses the Virtual Network. So they have to be deleted in that order. But they are pretty quick operations unlike the create/delete VM so I believe it should be fine.